api/v1alpha1/ndb_api_helpers.go

```diff
@@ -159,43 +160,151 @@ func GetNoneTimeMachineSLA(ctx context.Context, ndbclient *ndbclient.NDBClient)
+	//TODO: Do we need the explicit seggregation? Why can't we simply use allProfiles as that is a superset of profiles
+	genericProfiles := util.Filter(allProfiles, func(p ProfileResponse) bool { return p.EngineType == DATABASE_ENGINE_TYPE_GENERIC })
+	dbEngineSpecificProfiles := util.Filter(allProfiles, func(p ProfileResponse) bool { return p.EngineType == GetDatabaseEngineName(instanceSpec.Type) })
+
+	// Logic for fetching the Compute Profile
+	log.Info("Resolving the Compute profile using details:", "Profile", profiles.Compute)
 
-	// Specifying the usage of small compute profiles
-	computeProfiles := util.Filter(genericProfiles, func(p ProfileResponse) bool {
+	// TODO: Consider whether these util methods should go to a Util class
+	computeOOBProfileResolver := func(p ProfileResponse) bool {
 		return p.Type == PROFILE_TYPE_COMPUTE && strings.Contains(strings.ToLower(p.Name), "small")
-	})
+	}
+
+	computeProfile, err := profiles.Compute.resolve(ctx, allProfiles, computeOOBProfileResolver)
+	log.Info("Compute: ", "computeProfile", computeProfile)
+
+	if err != nil {
+		err = errors.New("error occurred while getting the compute profile details using the profile information")
+		log.Error(err, "Error:")
+		return nil, err
+	}
+
+	log.Info("Populating compute profile in the request...", "computeProfile", computeProfile)
+	profileMap[PROFILE_TYPE_COMPUTE] = computeProfile
+
+	// TODO: Do we really need storage profile?
+	// Logic for fetching the Storage Profile
 	storageProfiles := util.Filter(genericProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_STORAGE })
+	profileMap[PROFILE_TYPE_STORAGE] = storageProfiles[0]
+
+	// Logic for fetching the Software Profile
+	log.Info("Resolving the Software profile using details:", "Profile", profiles.Software)
+
 	// Specifying the usage of single instance topoplogy
-	softwareProfiles := util.Filter(dbEngineSpecificProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_SOFTWARE && p.Topology == TOPOLOGY_SINGLE })
-	networkProfiles := util.Filter(dbEngineSpecificProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_NETWORK })
-	dbParamProfiles := util.Filter(dbEngineSpecificProfiles, func(p ProfileResponse) bool { return p.Type == PROFILE_TYPE_DATABASE_PARAMETER })
+	softwareOOBProfileResolver := func(p ProfileResponse) bool {
+		return p.Type == PROFILE_TYPE_SOFTWARE && p.Topology == TOPOLOGY_SINGLE
+	}
 
-	// Some profile not found, return an error
-	if len(computeProfiles) == 0 || len(softwareProfiles) == 0 || len(storageProfiles) == 0 || len(networkProfiles) == 0 || len(dbParamProfiles) == 0 {
-		err = errors.New("oob profile: one or more OOB profile(s) were not found")
-		return
+	softwareProfile, err := profiles.Software.resolve(ctx, dbEngineSpecificProfiles, softwareOOBProfileResolver)
+	log.Info("Software:", "softwareProfile", softwareProfile)
```

Comment:
Can we restructure the code to have all the oobresolver methods definition together, followed by resolve calls (and error checking), followed by setting the values in the map? (We can also reduce logs for individual profiles and log them together in the end)

Reply:
To declutter, I have extracted out the resolver methods outside the GetProfiles method. I have also reduced the individual logs - I am only logging once for each type of profile inside the "resolve" method

I would still prefer to keep updating the profilesMap because in case we failed after the 1st resolution, then the final map we are printing at the end would be inconsistent.

Reply:
As discussed offline, failing would return directly. Resolving, thanks! :)